### PR TITLE
references.cc: clarify search optimization

### PR DESCRIPTION
**Pull request by milahu**
improve code readability

### CODE REVIEW

File: src/libstore/references.cc [hunk 1]
```diff
@@ -32,7 +32,7 @@ static void search(
         bool match = true;
         for (j = refLength - 1; j >= 0; --j)
             if (!isBase32[(unsigned char) s[i + j]]) {
-                i += j + 1;
+                i += j + 1; // skip refLength bytes
```
Comment (milahu) [inline src/libstore/references.cc:35]: more precise: `seek to next block`

**Comment by Ericson2314** [comment]
The original comment didn't really help me, so I added some more.

File: src/libutil/references.cc [hunk 2]
```diff
@@ -12,7 +12,10 @@
 namespace nix {
 
 
-static size_t refLength = 32; /* characters */
+/**
+ * A Nix reference is 32 characters long
```
Comment (roberth) [inline src/libutil/references.cc:16]: From what I know a "Nix"(?) reference is a relation between store objects, so this still makes no sense.
Is this about the store path hash part?

OT: why isn't this in libstore if it's about store concepts?

File: src/libutil/references.cc [hunk 3]
```diff
@@ -29,15 +32,21 @@ static void search(
     });
 
     for (size_t i = 0; i + refLength <= s.size(); ) {
```
Comment (roberth) [inline src/libutil/references.cc:34]:
```suggestion
    /* Nested loops, but the inner loop also advances i, so
       this is an asymptotically optimal solution for finding
       hash-like substrings. */
    for (size_t i = 0; i + refLength <= s.size(); ) {
```

It screams mutual recursion, but I'll ignore that.

File: src/libutil/references.cc [hunk 4]
```diff
@@ -29,15 +32,21 @@ static void search(
...
+           reference (because they are all base-32 characters). */
+        bool possibleReference = true;
+        for (size_t j = refLength; j > 0; --j)
+            if (!isBase32[(unsigned char) s[i + j - 1]]) {
+                /* We found an invalid character at index `i + j - 1`,
+                   the next possible reference (which doesn't contain
+                   it) starts at index `i + j`. */
+                i += j;
+                /* Use variable because we cannot directly continue in
+                   the outer loop in C++. */
```
Comment (roberth) [inline src/libutil/references.cc:45]: Edsger Dijkstra wouldn't mind.

**Review by roberth — CHANGES_REQUESTED** [review]

File: src/libutil/references.cc (hunk 2 above)
Comment (Ericson2314) [reply src/libutil/references.cc:16]: This happened in 3ebe1341abe1b0ad59bd4925517af18d9200f818 with my assistance. I got fooled by the types and thought this was store agnostic :(. I agree that given how the code is written today, it doesn't belong here.

Comment (Ericson2314) [reply src/libutil/references.cc:16]: @roberth I say we merge this after perhaps rephrasing the above, and then I'll move it back. How does that sound?

**Pull request closed without merging**